pageserver/src/tenant/storage_layer.rs

```diff
@@ -336,32 +312,328 @@ impl LayerAccessStats {
     ///
+    /// This might never come unless eviction called periodically.
+    status: tokio::sync::broadcast::Sender<Status>,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Status {
+    Evicted,
+    Downloaded,
+}
+
+impl std::fmt::Display for LayerE {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.layer_desc().short_id())
+    }
+}
+
+impl std::fmt::Debug for LayerE {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self)
+    }
+}
+
+impl AsLayerDesc for LayerE {
+    fn layer_desc(&self) -> &PersistentLayerDesc {
+        &self.desc
+    }
+}
+
+impl Drop for LayerE {
+    fn drop(&mut self) {
+        if !*self.wanted_garbage_collected.get_mut() {
+            // should we try to evict if the last wish was for eviction?
+            // feels like there's some hazard of overcrowding near shutdown near by, but we don't
+            // run drops during shutdown (yet)
+            return;
+        }
+
+        // TODO: spawn_blocking?
+        let span = tracing::info_span!(parent: None, "layer_drop", tenant_id = %self.layer_desc().tenant_id, timeline_id = %self.layer_desc().timeline_id, layer = %self);
+
+        // SEMITODO: yes, this is sync, could spawn as well..
+        let _g = span.entered();
+
+        let mut removed = false;
+        match std::fs::remove_file(&self.path) {
+            Ok(()) => {
+                removed = true;
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
+                // FIXME: unsure how to handle this. there are no deleted by mistake here, but it
+                // feels like the downloadedness state tracking, and so knowing here if the file
+                // should be present or not, requires load_layer_map fixing.
+            }
+            Err(e) => {
+                // FIXME: it is possible, that we've just evicted the layer or it was always remote
+                tracing::error!(layer = %self, "failed to remove garbage collected layer: {e}");
+            }
+        }
+
+        if let Some(timeline) = self.timeline.upgrade() {
+            if removed {
+                timeline
+                    .metrics
+                    .resident_physical_size_gauge
+                    .sub(self.layer_desc().file_size);
+            }
+            if let Some(remote_client) = timeline.remote_client.as_ref() {
+                let res =
+                    remote_client.schedule_layer_file_deletion(&[self.layer_desc().filename()]);
+
+                if let Err(e) = res {
+                    if !timeline.is_active() {
+                        // downgrade the warning to info maybe?
+                    }
+                    tracing::warn!(layer=%self, "scheduling deletion on drop failed: {e:#}");
+                }
+            }
+        } else {
+            // no need to nag that timeline is gone
```

Comment:
I cannot think of a case when. On shutdown for detach/ignore we don't clear the layermap before dropping the tenant.

`Tenant -> Timeline -> LayerManager -> LayerE` (-> keeps alive).

If this was a strong, I think it would always leak the Timeline, but perhaps the comment could be different, and log this suspected never to happen situation.

Reply:
Hmm actually I think that in the valid case of `remote_tenant_from_memory` we would get this warning. I'll need to test it out.

We would get this warning because timeline is already decided to be dropped, so you can no longer upgrade the Weak.

Reply:
I think this follows from what I mentioned earlier. Tenant will be ripped out of the global tenant map, the dropped. That allows timelines to be dropped, and automatic Timeline::drop the layers are dropped, and so we cannot upgrade anymore.

However, this situation would always create a NotFound error from deletion, so I will at least need to tune the comment on that arm.